### PR TITLE
Use WkbType instead of GeometryType for addPart

### LINKED ISSUE #57255: add part to geometryless feature does not work on single curvepolygon layer
### What is the bug or the crash?

add part map does not work for single curvepolygon layers when trying to add part to geometryless feature

### Steps to reproduce the issue

create a curvepolygon memory layer
create a geometryless feature from the attribute table
select it
try to add part

### Versions

at least 3.34+

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [ ] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Djedouas**
## Description

Deprecating `QgsGeometry::addPart` in favor of `QgsGeometry::addPartV2` to use `Qgis::WkbType` instead of `Qgis::GeometryType` because the latter can't handle curved geometries (not specific enough).

Fixes #57255

### CODE REVIEW

File: src/core/geometry/qgsgeometry.h [hunk 1]
```diff
@@ -908,24 +908,51 @@ class CORE_EXPORT QgsGeometry
...
      * \returns OperationResult a result code: success or reason of failure
+     * \deprecated since QGIS 3.38 - will be removed in QGIS 4.0. Use addPartV2 which accepts Qgis::WkbType geometry type instead of Qgis::GeometryType.
+     */
+    Q_DECL_DEPRECATED Qgis::GeometryOperationResult addPart( QgsAbstractGeometry *part SIP_TRANSFER, Qgis::GeometryType geomType = Qgis::GeometryType::Unknown ) SIP_DEPRECATED;
+
+    /**
+     * Adds a new part to this geometry.
+     * \param part part to add (ownership is transferred)
+     * \param wkbType default WKB type to create if no existing geometry
+     * \returns OperationResult a result code: success or reason of failure
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometry.h:953]: Add \since

File: src/core/geometry/qgsgeometry.h [hunk 2]
```diff
@@ -908,24 +908,51 @@ class CORE_EXPORT QgsGeometry
...
      * \returns OperationResult a result code: success or reason of failure
+     * \deprecated since QGIS 3.38 - will be removed in QGIS 4.0. Use addPartV2 which accepts Qgis::WkbType geometry type instead of Qgis::GeometryType.
+     */
+    Q_DECL_DEPRECATED Qgis::GeometryOperationResult addPart( const QgsPointSequence &points, Qgis::GeometryType geomType = Qgis::GeometryType::Unknown ) SIP_PYNAME( addPoints ) SIP_DEPRECATED;
+
+    /**
+     * Adds a new part to a the geometry.
+     * \param points points describing part to add
+     * \param wkbType default WKB type to create if no existing geometry
+     * \returns OperationResult a result code: success or reason of failure
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometry.h:936]: Add \since

File: src/core/geometry/qgsgeometry.h [hunk 3]
```diff
@@ -908,24 +908,51 @@ class CORE_EXPORT QgsGeometry
...
+     * \deprecated since QGIS 3.38 - will be removed in QGIS 4.0. Use addPartV2 which accepts Qgis::WkbType geometry type instead of Qgis::GeometryType.
      */
-    Qgis::GeometryOperationResult addPart( const QVector<QgsPointXY> &points, Qgis::GeometryType geomType = Qgis::GeometryType::Unknown ) SIP_PYNAME( addPointsXY );
+    Q_DECL_DEPRECATED Qgis::GeometryOperationResult addPart( const QVector<QgsPointXY> &points, Qgis::GeometryType geomType = Qgis::GeometryType::Unknown ) SIP_PYNAME( addPointsXY ) SIP_DEPRECATED;
+
+    /**
+     * Adds a new part to a the geometry.
+     * \param points points describing part to add
+     * \param wkbType default WKB type to create if no existing geometry
+     * \returns OperationResult a result code: success or reason of failure
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometry.h:919]: Since

File: src/core/geometry/qgsgeometry.cpp [hunk 4]
```diff
@@ -996,6 +1019,42 @@ Qgis::GeometryOperationResult QgsGeometry::addPart( QgsAbstractGeometry *part, Q
   return QgsGeometryEditUtils::addPart( d->geometry.get(), std::move( p ) );
 }
 
+Qgis::GeometryOperationResult QgsGeometry::addPartV2( QgsAbstractGeometry *part, Qgis::WkbType wkbType )
+{
+  std::unique_ptr< QgsAbstractGeometry > p( part );
+  if ( !d->geometry )
+  {
+    switch ( QgsWkbTypes::flatType( wkbType ) )
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometry.cpp:1027]: This method should also accept the multi types:

```suggestion
    switch ( QgsWkbTypes::singleType( QgsWkbTypes::flatType( wkbType ) ) )
```

**Review by nyalldawson — COMMENTED** [review]
Can you add unit tests for the QgsGeometry V2 methods as well?

File: src/core/geometry/qgsgeometry.cpp [hunk 5]
```diff
@@ -996,6 +1019,42 @@ Qgis::GeometryOperationResult QgsGeometry::addPart( QgsAbstractGeometry *part, Q
...
+        reset( nullptr );
+        return Qgis::GeometryOperationResult::AddPartNotMultiGeometry;
+    }
+  }
+  else
+  {
+    detach();
+  }
+
+  convertToMultiType();
```
Comment (3nids) [inline src/core/geometry/qgsgeometry.cpp:1054]: this can be moved to the else block, no?

Comment (Djedouas) [reply src/core/geometry/qgsgeometry.cpp:1056]: that's right, because if we get here and went through the switch then the geometry **is** multi type.

**Review by lbartoletti — APPROVED** [review]

**Pull request merged**